### ISSUE TITLE
Tracking issue for the 2023 revision of the array API specification

### ISSUE THREAD

**Issue by kgryte** [data-apis/array-api#643]
This is a tracking issue for tasks related to the 2023 revision of the array API specification. The issue is comprised of 3 parts: general todos, new APIs, and issues needing triage.

## TODO

- [x] Add page discussing delayed evaluation

    - ref: <https://github.com/data-apis/array-api/issues/642#issuecomment-1601386825>
    - PR: <https://github.com/data-apis/array-api/pull/652>
    - PR: <https://github.com/data-apis/array-api/pull/708>

- [x] Clarify out-of-bounds behavior in `take`

    - ref: <https://github.com/data-apis/array-api/issues/177#issuecomment-1513903987>
    - PR: <https://github.com/data-apis/array-api/pull/701>

- [x] Clarify minimum rank upper bound

    - ref: <https://github.com/data-apis/array-api/issues/479>
    - PR: <https://github.com/data-apis/array-api/pull/702>

- [x] Clarify sum description when `dtype=None`

    - ref: <https://github.com/data-apis/array-api/issues/623>
    - PR: <https://github.com/data-apis/array-api/pull/666>

- [x] Resolve choice of conjugation in complex inner product

    - ref: <https://github.com/data-apis/array-api/issues/616>
    - PR: <https://github.com/data-apis/array-api/pull/723>

- [x] Clarify broadcasting in linalg functions that accept an axis

    - ref: <https://github.com/data-apis/array-api/issues/617>

- [x] Fix descriptions for constants in rendered specification

    - ref: <https://github.com/data-apis/array-api/issues/601>
    - PR: <https://github.com/data-apis/array-api/pull/659>

- [x] Clarify index array in `unique_all`

    - ref: <https://github.com/data-apis/array-api/issues/588>
    - PR: <https://github.com/data-apis/array-api/pull/700>
    - PR: <https://github.com/data-apis/array-api/pull/752>

- ~~[ ] Add note clarifying `__imatmul__` broadcast behavior~~

    - ref: <https://github.com/data-apis/array-api/issues/509>
    - **Update**: moved to 2024

- ~~[ ] Add section indicating it's okay to duck type objects (finfo, constants)~~

    - ref: <https://github.com/data-apis/array-api/issues/405#issuecomment-1078263508>
    - **Update**: moved to 2024

- [x] DLPack related
   - **Anything more to discuss regarding DLPack and readonly/immutable arrays?** (priority: high)
      - ref: <https://github.com/data-apis/array-api/issues/191>
      - PR: <https://github.com/data-apis/array-api/pull/602>
   - **Anything more to say for forcing back to host?** (priority: high; include as part of DLPack discussion?)
      - ref: <https://github.com/data-apis/array-api/issues/626>

## New APIs

- [x] copysign

    - task size: XS
        - IEEE 754
        - risk: TensorFlow does not implement
    - priority: low
    - ref: <https://github.com/data-apis/array-api/issues/593>
    - PR: <https://github.com/data-apis/array-api/pull/693>

- ~~[ ] cbrt~~

    - task size: XS
        - IEEE 754
        - risk: need PyTorch to implement
    - priority: low
    - ref: <https://github.com/data-apis/array-api/issues/590>
    - **Update**: PyTorch still does not have a dedicated `cbrt` API. We can revisit in 2024.

- ~~[ ] exp2~~

    - task size: XS
        - IEEE 754
        - risk: MXNet does not implement
    - priority: low
    - ref: <https://github.com/data-apis/array-api/issues/591>
    - **Update**: this is low priority and should be revisited in 2024.

- [x] hypot

    - task size: XS
        - IEEE 754
        - risk: none (implemented across all major libraries)
    - priority: medium
    - ref: <https://github.com/data-apis/array-api/issues/544>
    - PR: <https://github.com/data-apis/array-api/pull/703>

- ~~[ ] copy (for copying data)~~

    - task size: S
        - risk: could become M depending on community feedback; need to ensure library alignment
    - priority: ???
    - ref: <https://github.com/data-apis/array-api/issues/495>
    - decisions:
        - name: copy vs clone
    - **Update**: this is low priority and not clear whether it's needed.

- [x] moveaxis

    - task size: S
    - priority: medium
        - no simple equivalent
    - ref: <https://github.com/data-apis/array-api/issues/483>
    - PR: <https://github.com/data-apis/array-api/pull/656>

- [x] searchsorted

    - task size: S
    - priority: medium
    - ref: <https://github.com/data-apis/array-api/issues/688>
    - PR: <https://github.com/data-apis/array-api/pull/699>

- ~~[ ] angle~~

    - task size: S
        - complex number dtypes
        - risk: MXNet does not implement
    - priority: low
    - ref: <https://github.com/data-apis/array-api/issues/595>
    - **Update**: this is low priority and should be revisited in 2024.

- ~~[ ] Hermitian transpose~~

    - task size: S
    - priority: low
        - convenience function
    - ref: <https://github.com/data-apis/array-api/issues/555>
    - **Update**: this is low enough priority that this can be bumped to the 2024 revision.

- [x] Inspection API namespace

    - task size: M
    - priority: high
    - ref: <https://github.com/data-apis/array-api/issues/640>
    - PR: <https://github.com/data-apis/array-api/pull/689>

- [x] clip

    - task size: M
        - bikeshed
    - priority: medium
    - ref: <https://github.com/data-apis/array-api/issues/482>
    - decisions:
        - type promotion
        - kwarg names
        - kwarg value types
    - PR: <https://github.com/data-apis/array-api/pull/715>

- [x] cumulative sum

    - task size: M
         - bikeshed and alignment
    - priority: medium
    - ref: <https://github.com/data-apis/array-api/issues/597>
    - decisions:
        - name
        - optional argument for initial value
    - PR: <https://github.com/data-apis/array-api/pull/653>

- ~~[ ] cumulative prod~~

    - task size: M
        - bikeshed and alignment
    - priority: low
    - ref: <https://github.com/data-apis/array-api/issues/598>
    - decisions:
        - name
        - optional argument for initial value
    - **Update**: this is low enough priority that it can wait until the 2024 specification revision.

- ~~[ ] topk~~

    - task size: M
    - priority: medium
    - ref: <https://github.com/data-apis/array-api/issues/629>
    - decisions:
        - name
    - PR: <https://github.com/data-apis/array-api/pull/722>
    - **Update**: moved to 2024 to allow for further R&D.

- [x] repeat/tile

    - task size: M
        - risk: potential to become L; need to ensure community alignment
    - priority: medium
        - xarray adoption
    - ref: <https://github.com/data-apis/array-api/issues/187#issuecomment-1554266113>
    - ref: <https://github.com/data-apis/array-api/issues/654>
    - ref: <https://github.com/data-apis/array-api/issues/655>
    - PR: <https://github.com/data-apis/array-api/pull/690>
    - PR: <https://github.com/data-apis/array-api/pull/692>

- ~~[ ] nan-reductions~~

    - task size: L
         - risk: requires NumPy buy-in
    - priority: high
    - ref: <https://github.com/data-apis/array-api/issues/621>
    - decisions:
        - API design
        - engagement with NumPy community
    - **Update**: this requires additional discussion and consensus and should be moved to the 2024 revision.

- ~~[ ] logsumexp~~

    - task size: L
        - risk: requires NumPy buy-in (would duplicate behavior present in SciPy)
    - priority: medium
    - ref: <https://github.com/data-apis/array-api/issues/596>
    - requires addition to NumPy
    - **Update**:
    	- this API was also mentioned as a candidate for inclusion in the special functions extension: https://github.com/data-apis/array-api/issues/725

## Triage

- Specification maintenance tasks

    - ref: <https://github.com/data-apis/array-api/issues/533>

- LU factorization

    - ref: <https://github.com/data-apis/array-api/issues/627>

- Anything more to say regarding libraries negotiating that they understand NumPy arrays?

    - ref: <https://github.com/data-apis/array-api/issues/586>

- Anything more to discuss for run-time version of API (static typing)?

    - ref: <https://github.com/data-apis/array-api/issues/584>

- Make dtypes obey Python hashing rules?

    - ref: <https://github.com/data-apis/array-api/issues/582>
    - requires changes to NumPy

- Specify that `__getitem__` and `__setitem__` accepts `SupportsIndex`

    - ref: <https://github.com/data-apis/array-api/issues/383>

- Special cases for +/-0 around branch cuts

    - ref: <https://github.com/data-apis/array-api/issues/570>

- Add batch support to `linalg.outer`

    - ref: <https://github.com/data-apis/array-api/issues/242>
    - awaiting PyTorch and updates to NumPy

- Support for closeness comparison

    - ref: <https://github.com/data-apis/array-api/issues/170>

- Enforcing shape to accommodate data-dependent (unknown) dimensionality

    - ref: <https://github.com/data-apis/array-api/issues/494>

- Adding spec guidance regarding exceptions to positional-only arguments

    - ref: <https://github.com/data-apis/array-api/issues/475>

**Comment by leofang** [comment]
Should we include #589 or push to 2024 (and perhaps backport it)?

**Comment by kgryte** [comment]
@leofang My inclination is that #589 should be pushed to 2024 with the possibility of backporting. I think it is lower priority compared to some of the other items we need to get over the finish line before cutting the 2023 release.

**Comment by leofang** [comment]
>   * ref: [to_device() -- any way to force back to host "portably?" #626](https://github.com/data-apis/array-api/issues/626)
> * Anything more to say regarding libraries negotiating that they understand NumPy arrays?

I raised #741 to address this.

**Comment by rgommers** [comment]
Everything in this tracking issue is now done (or moved to 2024, or descoped), so let's close it. On to the changelog and tagging a release. Thanks all!
